cirq-core/cirq/value/product_state.py

```diff
@@ -63,6 +64,9 @@ def __init__(self, states=None):
             # coverage: ignore
             states = dict()
 
+        n = len(states)
+        object.__setattr__(self, '_qid_shape', (2,) * n)
+        object.__setattr__(self, '_dim', 2**n)
```

Comment:
Does `super().__init__()` violate the `frozen=True` contract?

Reply:
Yes, I considered changing the superclass initializer to the above. That allowed use of `super.__init__()` here, but looked ugly and created "_qid_shape is not defined in init" warnings. This seemed like the best alternative.